### ISSUE TITLE
Understanding the consistency model during concurrent writes/read

### ISSUE THREAD

**Issue by nicola** [solid/solid#96]
Once with Tim's pad we broke ldnode since we had two concurrent writing and we have a bug, that we put on the side in September https://github.com/linkeddata/ldnode/issues/137.

What is our consistency model? In my opinion, this should not be on the spec, since every implementor can decide its own way of doing this, however, as implementors I think we should agree on.

- Do we want to give linearizability on read, write, append or write and append, or just on writes?
- Do we allow stale copies on read?
- Do we really need linearizability on append?

In my opinion the case to avoid are the following:

1. If U1 is reading R1, U2 shouldn't wait U1 to finish to read, append or write
2. If U1 appends to R1, U2 shouldn't wait U1 to finish to read or append.
3. If U1 writes R1, U2 shouldn't wait U1 to finish to read
4. (Bonus) If U1 writes R1, U2 shouldn't wait U1 to finish in order to write (allow concurrent write & only take the latest or notify a user they failed)

- For (1) to happen, we need no locks on read
- For (2) to happen, we need to have R1 be represented by multiple files (if you append concurrently, you may have the second append writing in the middle of the first append), unless we do some clever thing (like in GFS) where we compute an offset of the file where the two requests should write
- For (3) to happen, we need to put a write lock
- For (4) to happen, U1 writes on R1.back1 and U2 writes on R1.back2, when they are done the system copies R1.back* to R1 holding a write lock

I am not sure what the concurrency model is in `gold`, in `ldnode` is yet to be written

**Comment by akuckartz** [comment]
> In my opinion, this should not be on the spec, since every implementor can decide its own way of doing this ...

I do not understand this argument against specifying the concurrency behavior.

**Comment by nicola** [comment]
@akuckartz assume your database doesn't support the strategy that we choose, if you want your users to write and read and append one by one, having a global lock at each action, you should still be able to do so. (this is more of an opinion rather than a strong argument)

**Comment by sandhawke** [comment]
There are extensive discussions in LDP about this, mostly useless.

Your proposal loses me when you talk about waiting.

As I recall the key bit, and this is in HTTP, is to treat all HTTP operations as atomic.    That's trivial to implement for small resources,  under the os block size, and gets arbitrarily expensive for large resources - But one can probably avoid the cost by issuing error codes or blocking.

But as a model, doesn't that settle the issue?

**Comment by nicola** [comment]
If you treat all the operations as atomic, then you will have to wait for an operation to finish.
Imagine there is a 1GB file sitting in my pod and you update it. In order for me to read that document will I have to wait you to finish uploading it again or can I just read the one that is there without waiting for you.

In your atomic model, new operations needs the previous operation to complete for them to proceed. Instead, in the model that I suggest (with the examples, including 4), there is no need to wait before doing any operation.

**Comment by sandhawke** [comment]
No, they don't need to wait, they can be given the old version.    Or they can be given a try-again-later error.

**Comment by nicola** [comment]
@sandhawke in the first case, then the HTTP request is not atomic and we agree on my scheme, in the second case they are atomic and we have two different schemes

**Comment by sandhawke** [comment]
To be clear, I can't understand the language of your proposal.

Atomic does not speak to ordering.

Client1 starts a PUT

While it's streaming the new data.... Client2 begins a GET

The server can either:
1. Give some kind of try-again-later error
2. Keep the connection open but not return data to client2 until client1 is done
3.Give client2 a copy of the data as it was before client1's PUT began

All of these options are okay according to the specs, as I recall.    3 is still atomic, and linear (consistent) if you think of the PUT as occurring at the end (which is quite reasonable).

**Comment by nicola** [comment]
@sandhawke I am sorry, when I meant atomic, I was picturing a timeline of operations

1. `| Alice POST R | Bob GET R` (the atomic case)
2. `| Alice starts POST R | Bob starts GET R | Bob finish GET R | A finish POST R` (non atomic case)

I absolutely agree that the spec allows any of this to happen (this is why I don't want to add this on the spec), but I want to discuss which one `ldnode`, for example, should implement. Your three examples are the same ones I am discussing (I also add some arguments on append). The scheme I suggest (my points 1-4), I argue is the one implementors should aim at, since no request is blocking.

**Comment by sandhawke** [comment]
The difference between atomic and non-atomic is whether anyone can see the data partly changed.    If PUT replaces the resource with content "aa" with the content "bb", the operation is atomic if every GET returns either "aa" or "  bb".     To be non-atomic the server would have to return "ab" or "ba".   Of course with a two byte file it would be hard to be non atomic.     But with a 2GB file it would be easy.     The easy way to make it atomic in the current filesystem implementation is for PUT to stream to a temp file then rename after writing is done.   PATCH will require copying the data before patching.

**Comment by sandhawke** [comment]
There should of course be a test for this in the client tests.    :-)

**Comment by nicola** [comment]
(in my example I consider the database/fs writes to be atomic, but not the HTTP request+LDP write+HTTP response)
